resources/ManageAPICredentials.php

```diff
@@ -0,0 +1,305 @@
+        // Remove the credentials that would have been deleted if not in dry-run mode
+        $creds = array_udiff($creds, $deletedCreds, '\gocdb\scripts\compareCredIds');
+
+        reportDryRun($creds, "sending warning emails");
+        return $creds;
+    }
+
+    /* @var $api APIAuthentication */
+    foreach ($creds as $api) {
+        $lastUsed = $api->getLastUseTime();
+
+        $timeDiff = (new DateTime())->diff($lastUsed);
+        $lastUseMonths = $timeDiff->format('%m');
+
+        sendWarningEmail($fromEmail, $replyToEmail, $api, intval($lastUseMonths), $deletionThreshold);
+    }
+
+    return $creds;
+}
+/**
+ * Find API credentials unused for a number of months.
+ *
+ * Find API credentials which have not been used for a number of months prior to a given base time based
+ * on the credential property lastUseTime.
+ *
+ * @param \Doctrine\Orm\EntityManager $entitymanager A valid Doctrine Entity Manager
+ * @param \DateTime     $baseTime       Time from which interval of no-use is measured
+ * @param int           $threshold      The number of months of no-use prior to $baseTime to use for selection
+ */
+function getCreds($entityManager, $baseTime, $threshold)
+{
+    $qbl = $entityManager->createQueryBuilder();
+
+    $qbl->select('cred')
+        ->from('APIAuthentication', 'cred')
+        ->where('cred.lastUseTime < :threshold');
+
+    $timeThresh = clone $baseTime;
+
+    if ($threshold > 0) {
+        $timeThresh->sub(new DateInterval("P" . $threshold . "M"));
+    }
+
+    $qbl->setParameter('threshold', $timeThresh->format('Y-m-d 00:00:00'));
+
+    $creds = $qbl->getQuery()->getResult();
+
+    return $creds;
+}
+
+/**
+ * Format and send warning emails.
+ *
+ * Send emails to API credential owner and the registered site address warning of impending credential deletion
+ * if the credential remains unused until a given threshold of months.
+ *
+ * @param string    $fromEmail          Email address to use as sender's (From:) address
+ * @param string    $replyToEmail       Email address for replies (Reply-To:)
+ * @param \APIAuthentication $api       Credential to warn about
+ * @param int       $elapsedMonths      The number of months of non-use so far.
+ * @param int       $deleteionThreshold The number of months of no-use which will trigger deletion if reached.
+ * @return void
+ */
+function sendWarningEmail(
+    $fromEmail,
+    $replyToEmail,
+    \APIAuthentication $api,
+    $elapsedMonths,
+    $deletionThreshold
+) {
+    $user = $api->getUser();
+    $userEmail = $user->getEmail();
+    $siteName = $api->getParentSite()->getShortName();
+    $siteEmail = $siteName . ' <' . $api->getParentSite()->getEmail() . '>';
+
+    // Some code useful for testing
+    /*
+    print("\n!!!DUMMY SITE EMAIL!!! was $siteEmail\n\n");
+    $siteEmail = 'Name <name@address.com>';
+    */
+    /*
+    print("\n!!!DUMMY USER EMAIL!!! was $userEmail\n\n");
+    $userEmail = 'AnotherName <anothername@address.com>';
+    */
```

Comment:
I think it would be better for testing to use a dummy data set - so lets remove this commented out code?

Reply:
OK